agent/consul/state/peering.go

```diff
@@ -251,45 +251,42 @@ func (s *Store) peeringSecretsWriteTxn(tx WriteTxn, secret *pbpeering.PeeringSec
 
 	var toDelete []string
 	if existing != nil {
-		// Merge in existing stream secrets when persisting a new establishment secret.
-		// This is to avoid invalidating stream secrets when a new peering token
-		// is generated.
-		//
-		// We purposely DO NOT do the reverse of inheriting an existing establishment secret.
-		// When exchanging establishment secrets for stream secrets, we invalidate the
-		// establishment secret by deleting it.
-		if secret.GetEstablishment() != nil && secret.GetStream() == nil && existing.GetStream() != nil {
-			secret.Stream = existing.Stream
-		}
-
 		// Collect any overwritten UUIDs for deletion.
-		//
-		// Old establishment secret ID are always cleaned up when they don't match.
-		// They will either be replaced by a new one or deleted in the secret exchange RPC.
-		existingEstablishment := existing.GetEstablishment().GetSecretID()
-		if existingEstablishment != "" && existingEstablishment != secret.GetEstablishment().GetSecretID() {
-			toDelete = append(toDelete, existingEstablishment)
-		}
+		switch req.Operation {
+		case pbpeering.PeeringSecretsWriteRequest_OPERATION_GENERATETOKEN:
+			// Merge in existing stream secrets when persisting a new establishment secret.
+			// This is to avoid invalidating stream secrets when a new peering token
+			// is generated.
+			secret.Stream = existing.GetStream()
+
+			// When a new token is generated we replace any un-used establishment secrets.
+			if existingEstablishment := existing.GetEstablishment().GetSecretID(); existingEstablishment != "" {
+				toDelete = append(toDelete, existingEstablishment)
+			}
 
-		// Old active secret IDs are always cleaned up when they don't match.
-		// They are only ever replaced when promoting a pending secret ID.
-		existingActive := existing.GetStream().GetActiveSecretID()
-		if existingActive != "" && existingActive != secret.GetStream().GetActiveSecretID() {
-			toDelete = append(toDelete, existingActive)
-		}
+		case pbpeering.PeeringSecretsWriteRequest_OPERATION_EXCHANGESECRET:
+			// Avoid invalidating existing active secrets when exchanging establishment secret for pending.
+			secret.Stream.ActiveSecretID = existing.GetStream().GetActiveSecretID()
+
+			// When exchanging an establishment secret we invalidate the existing establishment secret.
+			if existingEstablishment := existing.GetEstablishment().GetSecretID(); existingEstablishment != "" {
```

Comment:
Should we error out if the establishment secret isn't set?

I am thinking that ExchangeSecret is called in quick succession and that we end up sending two raft logs with different pending secrets through. The first has already removed the secret and the next potentially is going to overwrite the pending one. Its super racey but I wonder if moving more validation logic into the state store like this would be helpful to mitigate those races.

Reply:
I put in a check to make sure it's set, but I think I'm actually going to pass in the establishment secret that was validated against to guard against the case where it was non-empty but different